allennlp/training/trainer.py

```diff
@@ -231,23 +441,49 @@ def _train_epoch(self, epoch: int) -> dict:
             train_generator_tqdm.set_description(description, refresh=False)
 
             # Log parameter values to Tensorboard
-            batch_num_total = num_training_batches * epoch + batch_num
             if batch_num_total % self._summary_interval == 0:
                 for name, param in self._model.named_parameters():
                     self._tensorboard.add_train_scalar("parameter_mean/" + name,
                                                        param.data.mean(),
                                                        batch_num_total)
                     self._tensorboard.add_train_scalar("parameter_std/" + name, param.data.std(), batch_num_total)
                     if param.grad is not None:
+                        if is_sparse(param.grad):
+                            # pylint: disable=protected-access
+                            grad_data = param.grad.data._values()
+                        else:
+                            grad_data = param.grad.data
                         self._tensorboard.add_train_scalar("gradient_mean/" + name,
-                                                           param.grad.data.mean(),
+                                                           grad_data.mean(),
                                                            batch_num_total)
                         self._tensorboard.add_train_scalar("gradient_std/" + name,
-                                                           param.grad.data.std(),
+                                                           grad_data.std(),
                                                            batch_num_total)
                 self._tensorboard.add_train_scalar("loss/loss_train", metrics["loss"], batch_num_total)
                 self._metrics_to_tensorboard(batch_num_total,
                                              {"epoch_metrics/" + k: v for k, v in metrics.items()})
+                # norm of gradients
+                if self._batch_grad_norm is not None:
+                    self._tensorboard.add_train_scalar("gradient_norm",
+                                                       self._batch_grad_norm,
+                                                       batch_num_total)
+
+            # Histogram logging
+            if self._should_log_histogram:
+                for name, param in self._model.named_parameters():
+                    if name in histogram_parameters:
+                        self._tensorboard.add_train_histogram("parameter_histogram/" + name,
+                                                              param,
+                                                              batch_num_total)
+
+            # Save model if needed.
+            if self._model_save_interval is not None and (
+                    time.time() - last_save_time > self._model_save_interval
+            ):
+                last_save_time = time.time()
```

Comment:
It would be better if you formatted this as a date - the current way is not great because it just looks like a float value and has no real meaning.

Reply:
I'd really not do that. These are easy to produce and parse in the formatted files (we need to parse the file name back into a computable form to determine the latest checkpoint).  It adds some additional complexity to make them more human readable.

Reply:
Why is having a regex for what you currently have harder than having a regex for a date? Seems strictly more informative and only a tiny bit of work to change. You can just load the date string in the same format and compare using `datetime` - parsing: `time.strptime(date1, "%d/%m/%Y")`